File: client/packages/app/lib/src/permissions.dart
// SPDX-License-Identifier: LicenseRef-PolyForm-Noncommercial-1.0.0
/// A client-side mirror of `crates/slimm-server/src/permissions.rs`.
///
/// Every bit here must match the server's `Permissions` constants exactly:
/// this is what `Me.permissions` and `Role.permissions` are bitmasks of. Used
/// only to decide what to show; every write is still re-authorized
/// server-side from scratch regardless of what a caller can see.
library;

abstract final class Perm {
  static const int administrator = 1 << 0;
  static const int viewChannel = 1 << 1;
  static const int sendMessages = 1 << 2;
  static const int manageMessages = 1 << 3;
  static const int manageChannels = 1 << 4;
  static const int manageRoles = 1 << 5;
  static const int kickMembers = 1 << 6;
  static const int banMembers = 1 << 7;
  static const int createInvite = 1 << 8;
  static const int addReactions = 1 << 9;
  static const int attachFiles = 1 << 10;
  static const int connect = 1 << 11;
  static const int speak = 1 << 12;
  static const int useCanvas = 1 << 13;
  static const int manageCanvas = 1 << 14;
  static const int manageServer = 1 << 15;

  /// Whether `@everyone`/`@here` actually wake anyone the caller mentions
  /// with them, rather than sitting as plain text; see
  /// `crates/slimm-server/src/permissions.rs`'s own doc on this bit.
  static const int mentionEveryone = 1 << 16;

  /// Run a fenced code block through this deployment's configured code
  /// runner, when one is configured at all (decision 0026). Defaults to
  /// nobody on every fresh deployment and DM; see the server's own doc on
  /// this bit for why.
  static const int runCode = 1 << 17;

  /// Read moderation history without `manageMessages`'s write power.
  static const int viewModerationHistory = 1 << 18;

  /// Every bit that has a name, in the fixed order the editor lists them.
  static const List<(int bit, String label)> editable = [
    (administrator, 'Administrator'),
    (viewChannel, 'View channels'),
    (sendMessages, 'Send messages'),
    (manageMessages, 'Manage messages'),
    (manageChannels, 'Manage channels'),
    (manageRoles, 'Manage roles'),
    (kickMembers, 'Kick members'),
    (banMembers, 'Ban members'),
    (createInvite, 'Create invites'),
    (addReactions, 'Add reactions'),
    (attachFiles, 'Attach files'),
    (connect, 'Join voice channels'),
    (speak, 'Speak in voice channels'),
    (useCanvas, 'Use the voice canvas'),
    (manageCanvas, 'Manage the voice canvas'),
    (manageServer, 'Manage Space settings'),
    (mentionEveryone, 'Mention @everyone and @here'),
    (runCode, 'Run code blocks'),
    (viewModerationHistory, 'View moderation history'),
  ];

  /// [editable] minus [administrator]: the server's evaluator returns every
  /// permission the moment a caller's roles carry [administrator], before it
  /// ever looks at a channel overwrite, so an allow or deny of it in a
  /// per-channel overwrite can never change anything in either direction.
  static final List<(int bit, String label)> channelOverwriteEditable =
      List.unmodifiable(editable.where((p) => p.$1 != administrator));

  /// [groups] flattened, in the same order - the channel permissions grid's
  /// row list, since a grid has no room for group headers of its own and
  /// [administrator] never applies to a per-channel overwrite (see
  /// [channelOverwriteEditable]).
  static final List<PermSpec> gridRows = List.unmodifiable([
    for (final group in groups) ...group.permissions,
  ]);

  /// Every grantable permission but [administrator], grouped and described
  /// the way the roles pane's Permissions tab lists them. Administrator is
  /// its own special row above every group, not a member of one, since it
  /// is not "one permission among others" - it grants all of them.
  static const List<PermGroup> groups = [
    PermGroup('Messages', [
      PermSpec(
        sendMessages,
        'Send messages',
        'Post in text channels this role can see.',
      ),
      PermSpec(
        manageMessages,
        'Manage messages',
        "Delete and pin other people's messages.",
      ),
      PermSpec(attachFiles, 'Attach files', 'Upload images and files.'),
      PermSpec(addReactions, 'Add reactions', 'React to messages with emoji.'),
    ]),
    PermGroup('Moderation', [
      PermSpec(
        manageRoles,
        'Manage roles',
        'Edit and assign roles below this one.',
        elevated: true,
      ),
      PermSpec(
        kickMembers,
        'Kick members',
        'Remove someone; they can rejoin with an invite.',
        elevated: true,
      ),
      PermSpec(
        banMembers,
        'Ban members',
        'Remove someone and block them from rejoining.',
        elevated: true,
      ),
      PermSpec(
        viewModerationHistory,
        'View moderation history',
        'Read past moderation actions, without the power to take new ones.',
      ),
    ]),
    PermGroup('Voice & canvas', [
      PermSpec(connect, 'Join voice channels', 'Connect to a voice channel.'),
      PermSpec(
        speak,
        'Speak in voice channels',
        'Transmit audio after joining.',
      ),
      PermSpec(
        useCanvas,
        'Use the voice canvas',
        'Draw and place objects during a call.',
      ),
      PermSpec(
        manageCanvas,
        'Manage the voice canvas',
        'Clear or reset the canvas for everyone in the call.',
      ),
    ]),
    PermGroup('Channels & invites', [
      PermSpec(
        viewChannel,
        'View channels',
        'See a channel this role is not otherwise denied.',
      ),
      PermSpec(
        manageChannels,
        'Manage channels',
        'Create, rename, reorder and delete channels.',
        elevated: true,
      ),
      PermSpec(
        createInvite,
        'Create invites',
        'Generate invite links to this Space.',
      ),
    ]),
    PermGroup('Space', [
      PermSpec(
        manageServer,
        'Manage Space settings',
        'Change deployment-wide settings for everyone.',
        elevated: true,
      ),
      PermSpec(
        mentionEveryone,
        'Mention @everyone and @here',
        'Wake every member with a mention.',
        elevated: true,
      ),
    ]),
    PermGroup('Extensions', [
      PermSpec(
        runCode,
        'Run code blocks',
        "Execute a fenced code block through this deployment's configured runner.",
      ),
    ]),
  ];
}

/// One permission's display metadata: the label and one-line description the
/// roles pane's Permissions tab and the channel permissions grid both read
/// off, plus whether it is tagged `elevated` - a permission that can act on
/// other members or the whole deployment, called out the same way the design
/// review does, not a bit the server treats specially.
class PermSpec {
  const PermSpec(
    this.bit,
    this.label,
    this.description, {
    this.elevated = false,
  });

  final int bit;
  final String label;
  final String description;
  final bool elevated;
}

/// A named run of [PermSpec]s, the roles pane's group headers
/// (`MESSAGES`, `MODERATION`, ...).
class PermGroup {
  const PermGroup(this.title, this.permissions);

  final String title;
  final List<PermSpec> permissions;
}

/// Whether a raw permission bitmask contains every bit in [required].
/// Mirrors the server's own `contains`: [administrator] is not special-cased
/// here, matching `Me.permissions` and `Role.permissions`, which already
/// carry it as an ordinary bit a caller can check for directly.
extension PermissionCheck on int {
  bool hasPermission(int required) => (this & required) == required;
}

File: crates/slimm-server/src/permissions.rs
// SPDX-License-Identifier: AGPL-3.0-only
//! The deny-by-default permission model and its evaluator.
//!
//! A permission set is a bitmask stored in a 63-bit `INTEGER`. Nothing is
//! granted implicitly: a user can only do what the evaluated set contains.
//!
//! [`evaluate`] resolves the effective set for a user in a channel, in this
//! fixed order:
//!
//! 1. Base. The `@everyone` role's permissions, then the union of every role the
//!    user holds.
//! 2. Administrator bypass. If the base set carries [`Permissions::ADMINISTRATOR`]
//!    the user gets everything and channel overwrites are skipped entirely.
//! 3. The `@everyone` channel overwrite: clear its deny bits, then set its allow.
//! 4. The role channel overwrites, aggregated across the user's roles with deny
//!    winning: set the union of allows, then clear the union of denies.
//! 5. The member channel overwrite, which is absolute and has the final say:
//!    clear its deny bits, then set its allow.
//!
//! Step 4 deviates from the allow-wins convention some platforms use: here a
//! deny on any of the user's roles wins over an allow on another, which is the
//! more conservative reading and matches the project's deny-by-default stance.
//! The member overwrite in step 5 is what re-grants a permission to a specific
//! user despite a role-level deny.

/// A set of permission bits.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, sqlx::Type)]
#[sqlx(transparent)]
pub struct Permissions(i64);

impl Permissions {
    /// No permissions. The starting point for every evaluation.
    pub const NONE: Self = Self(0);

    /// Grants every permission and bypasses all channel overwrites.
    pub const ADMINISTRATOR: Self = Self(1 << 0);
    /// See a channel and its history.
    pub const VIEW_CHANNEL: Self = Self(1 << 1);
    /// Post messages in a channel.
    pub const SEND_MESSAGES: Self = Self(1 << 2);
    /// Edit or delete other members' messages.
    pub const MANAGE_MESSAGES: Self = Self(1 << 3);
    /// Create, edit, and delete channels.
    pub const MANAGE_CHANNELS: Self = Self(1 << 4);
    /// Create, edit, and delete roles and their assignments.
    pub const MANAGE_ROLES: Self = Self(1 << 5);
    /// Remove members from the community.
    pub const KICK_MEMBERS: Self = Self(1 << 6);
    /// Ban members from the community.
    pub const BAN_MEMBERS: Self = Self(1 << 7);
    /// Create invites.
    pub const CREATE_INVITE: Self = Self(1 << 8);
    /// Add reactions to messages.
    pub const ADD_REACTIONS: Self = Self(1 << 9);
    /// Attach files to messages.
    pub const ATTACH_FILES: Self = Self(1 << 10);
    /// Join a voice channel.
    pub const CONNECT: Self = Self(1 << 11);
    /// Transmit audio in a voice channel.
    pub const SPEAK: Self = Self(1 << 12);
    /// View and draw on the voice canvas.
    pub const USE_CANVAS: Self = Self(1 << 13);
    /// Moderate canvas objects (move or remove others' work).
    pub const MANAGE_CANVAS: Self = Self(1 << 14);
    /// Change community-wide settings.
    pub const MANAGE_SERVER: Self = Self(1 << 15);
    /// Have an `@everyone` or `@here` mention actually wake anyone, rather
    /// than sit as plain text nobody is pushed for. See
    /// `push::recipients::resolved_mentions`, the one place this is read;
    /// `@everyone`/`@here` remain typeable without it, they just reach
    /// nobody extra, the same forgiving shape an unmatched `@nobody` has.
    pub const MENTION_EVERYONE: Self = Self(1 << 16);

    /// The union of every defined permission. What administrator resolves to.
    pub const ALL: Self = Self(
        Self::ADMINISTRATOR.0
            | Self::VIEW_CHANNEL.0
            | Self::SEND_MESSAGES.0
            | Self::MANAGE_MESSAGES.0
            | Self::MANAGE_CHANNELS.0
            | Self::MANAGE_ROLES.0
            | Self::KICK_MEMBERS.0
            | Self::BAN_MEMBERS.0
            | Self::CREATE_INVITE.0
            | Self::ADD_REACTIONS.0
            | Self::ATTACH_FILES.0
            | Self::CONNECT.0
            | Self::SPEAK.0
            | Self::USE_CANVAS.0
            | Self::MANAGE_CANVAS.0
            | Self::MANAGE_SERVER.0
            | Self::MENTION_EVERYONE.0,
    );

    /// Wraps a raw bitmask, for example one loaded from the database.
    pub const fn from_bits(bits: i64) -> Self {
        Self(bits)
    }

    /// The raw bitmask, for storage or the wire.
    pub const fn bits(self) -> i64 {
        self.0
    }

    /// Whether every bit in `other` is present.
    pub const fn contains(self, other: Self) -> bool {
        (self.0 & other.0) == other.0
    }

    /// Whether any bit in `other` is present.
    pub const fn intersects(self, other: Self) -> bool {
        (self.0 & other.0) != 0
    }

    /// The union of two sets.
    pub const fn union(self, other: Self) -> Self {
        Self(self.0 | other.0)
    }

    /// This set with `other`'s bits cleared.
    pub const fn remove(self, other: Self) -> Self {
        Self(self.0 & !other.0)
    }
}

/// A channel overwrite: bits to force on (`allow`) and bits to force off
/// (`deny`) for one role or one member.
#[derive(Debug, Clone, Copy)]
pub struct Overwrite {
    pub allow: Permissions,
    pub deny: Permissions,
}

/// Resolves the effective permissions for a user in a channel. See the module
/// docs for the precedence rules. Pass `None`/empty for a guild-level (no
/// channel) evaluation.
pub fn evaluate(
    everyone: Permissions,
    member_roles: &[Permissions],
    everyone_overwrite: Option<Overwrite>,
    role_overwrites: &[Overwrite],
    member_overwrite: Option<Overwrite>,
) -> Permissions {
    let mut base = everyone;
    for role in member_roles {
        base = base.union(*role);
    }

    if base.contains(Permissions::ADMINISTRATOR) {
        return Permissions::ALL;
    }

    if let Some(ow) = everyone_overwrite {
        base = base.remove(ow.deny).union(ow.allow);
    }

    let mut allow = Permissions::NONE;
    let mut deny = Permissions::NONE;
    for ow in role_overwrites {
        allow = allow.union(ow.allow);
        deny = deny.union(ow.deny);
    }
    base = base.union(allow).remove(deny);

    if let Some(ow) = member_overwrite {
        base = base.remove(ow.deny).union(ow.allow);
    }

    base
}

/// Masks a bitmask to all-zero unless it carries VIEW_CHANNEL.
///
/// A real channel the caller cannot view still passes its unrelated base
/// bits straight through unless an overwrite happens to deny them, and
/// `@everyone` usually grants something - so left unmasked, this would
/// answer differently for "channel does not exist" (forced NONE) than for
/// "channel exists, caller cannot view it, but nothing denies the bits
/// their base already grants", turning a caller into a channel-existence
/// oracle. Shared by `GET /channels/{channelId}/permissions` and
/// `Store::permissions_in_channels`; see
/// docs/decisions/0011-per-channel-permissions.md.
pub fn mask_unless_viewable(permissions: Permissions) -> Permissions {
    if permissions.contains(Permissions::VIEW_CHANNEL) {
        permissions
    } else {
        Permissions::NONE
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const VIEW: Permissions = Permissions::VIEW_CHANNEL;
    const SEND: Permissions = Permissions::SEND_MESSAGES;

    fn allow(p: Permissions) -> Overwrite {
        Overwrite {
            allow: p,
            deny: Permissions::NONE,
        }
    }

    fn deny(p: Permissions) -> Overwrite {
        Overwrite {
            allow: Permissions::NONE,
            deny: p,
        }
    }

    /// The masking that stops a channel being an existence oracle (decision
    /// 0011): without VIEW_CHANNEL every other bit is hidden, so a caller who
    /// cannot see the channel cannot read one permission off it; with it, the
    /// set passes through unchanged.
    #[test]
    fn mask_unless_viewable_hides_everything_without_view() {
        // A rich set still collapses to nothing when VIEW_CHANNEL is absent.
        let rich = SEND.union(Permissions::MANAGE_MESSAGES);
        assert_eq!(mask_unless_viewable(rich), Permissions::NONE);
        assert_eq!(mask_unless_viewable(Permissions::NONE), Permissions::NONE);
        // With VIEW_CHANNEL present, the whole set is visible unchanged.
        let viewable = VIEW.union(SEND);
        assert_eq!(mask_unless_viewable(viewable), viewable);
        assert_eq!(mask_unless_viewable(VIEW), VIEW);
    }

    #[test]
    fn nothing_is_granted_by_default() {
        let result = evaluate(Permissions::NONE, &[], None, &[], None);
        assert_eq!(result, Permissions::NONE);
        assert!(!result.contains(SEND));
    }

    #[test]
    fn everyone_base_and_role_union_accumulate() {
        let everyone = VIEW.union(SEND);
        let mod_role = Permissions::MANAGE_MESSAGES;
        let result = evaluate(everyone, &[mod_role], None, &[], None);
        assert!(result.contains(VIEW));
        assert!(result.contains(SEND));
        assert!(result.contains(Permissions::MANAGE_MESSAGES));
    }

    #[test]
    fn administrator_grants_all_and_ignores_denies() {
        let admin = Permissions::ADMINISTRATOR;
        // Even a member overwrite denying everything cannot touch an admin.
        let result = evaluate(
            Permissions::NONE,
            &[admin],
            None,
            &[deny(Permissions::ALL)],
            Some(deny(Permissions::ALL)),
        );
        assert_eq!(result, Permissions::ALL);
        assert!(result.contains(Permissions::BAN_MEMBERS));
    }

    #[test]
    fn everyone_overwrite_can_remove_a_base_permission() {
        let everyone = VIEW.union(SEND);
        let result = evaluate(everyone, &[], Some(deny(SEND)), &[], None);
        assert!(result.contains(VIEW));
        assert!(!result.contains(SEND));
    }

    #[test]
    fn role_overwrite_deny_wins_over_role_overwrite_allow() {
        // One role's overwrite allows SEND, another denies it. Deny wins.
        let everyone = VIEW;
        let result = evaluate(everyone, &[], None, &[allow(SEND), deny(SEND)], None);
        assert!(result.contains(VIEW));
        assert!(
            !result.contains(SEND),
            "a role-level deny wins over an allow"
        );
    }

    #[test]
    fn member_overwrite_regrants_over_a_role_deny() {
        // Role tier denies SEND; the member overwrite grants it back absolutely.
        let everyone = VIEW;
        let result = evaluate(everyone, &[], None, &[deny(SEND)], Some(allow(SEND)));
        assert!(result.contains(SEND), "member overwrite is absolute");
    }

    #[test]
    fn member_overwrite_deny_removes_a_granted_permission() {
        let everyone = VIEW.union(SEND);
        let result = evaluate(everyone, &[], None, &[], Some(deny(SEND)));
        assert!(result.contains(VIEW));
        assert!(!result.contains(SEND));
    }

    /// A bit omitted from `ALL` is a bit administrators do not hold, that the
    /// API refuses to grant, and that nothing else here would catch.
    #[test]
    fn all_is_the_union_of_every_named_permission() {
        let named = [
            Permissions::ADMINISTRATOR,
            Permissions::VIEW_CHANNEL,
            Permissions::SEND_MESSAGES,
            Permissions::MANAGE_MESSAGES,
            Permissions::MANAGE_CHANNELS,
            Permissions::MANAGE_ROLES,
            Permissions::KICK_MEMBERS,
            Permissions::BAN_MEMBERS,
            Permissions::CREATE_INVITE,
            Permissions::ADD_REACTIONS,
            Permissions::ATTACH_FILES,
            Permissions::CONNECT,
            Permissions::SPEAK,
            Permissions::USE_CANVAS,
            Permissions::MANAGE_CANVAS,
            Permissions::MANAGE_SERVER,
            Permissions::MENTION_EVERYONE,
        ];
        let union = named
            .into_iter()
            .fold(Permissions::NONE, Permissions::union);
        assert_eq!(union, Permissions::ALL);
    }

    #[test]
    fn contains_and_intersects_are_distinct() {
        let set = VIEW.union(SEND);
        assert!(set.contains(VIEW));
        assert!(set.contains(VIEW.union(SEND)));
        assert!(!set.contains(VIEW.union(Permissions::BAN_MEMBERS)));
        assert!(set.intersects(VIEW.union(Permissions::BAN_MEMBERS)));
        assert!(!set.intersects(Permissions::BAN_MEMBERS));
    }
}
